flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/casting/TimestampToTimestampCastRule.java

```diff
@@ -0,0 +1,98 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.planner.functions.casting;
+
+import org.apache.flink.table.data.TimestampData;
+import org.apache.flink.table.planner.codegen.calls.BuiltInMethods;
+import org.apache.flink.table.types.logical.LogicalType;
+import org.apache.flink.table.types.logical.LogicalTypeRoot;
+import org.apache.flink.table.types.logical.TimestampKind;
+import org.apache.flink.table.types.logical.TimestampType;
+import org.apache.flink.table.types.logical.utils.LogicalTypeChecks;
+
+import static org.apache.flink.table.planner.functions.casting.CastRuleUtils.staticCall;
+
+/**
+ * {@link LogicalTypeRoot#TIMESTAMP_WITHOUT_TIME_ZONE}/{@link
+ * LogicalTypeRoot#TIMESTAMP_WITH_LOCAL_TIME_ZONE} to {@link
+ * LogicalTypeRoot#TIMESTAMP_WITHOUT_TIME_ZONE}/{@link LogicalTypeRoot#TIMESTAMP_WITHOUT_TIME_ZONE}
+ * cast rule. Check and adjust if there is the precision changes.
+ */
+class TimestampToTimestampCastRule
+        extends AbstractExpressionCodeGeneratorCastRule<TimestampData, TimestampData> {
+
+    static final TimestampToTimestampCastRule INSTANCE = new TimestampToTimestampCastRule();
+
+    private TimestampToTimestampCastRule() {
+        super(
+                CastRulePredicate.builder()
+                        .input(LogicalTypeRoot.TIMESTAMP_WITHOUT_TIME_ZONE)
+                        .input(LogicalTypeRoot.TIMESTAMP_WITH_LOCAL_TIME_ZONE)
+                        .target(LogicalTypeRoot.TIMESTAMP_WITHOUT_TIME_ZONE)
+                        .target(LogicalTypeRoot.TIMESTAMP_WITH_LOCAL_TIME_ZONE)
+                        .build());
+    }
+
+    @Override
+    public String generateExpression(
+            CodeGeneratorCastRule.Context context,
+            String inputTerm,
+            LogicalType inputLogicalType,
+            LogicalType targetLogicalType) {
+        final int inputPrecision = LogicalTypeChecks.getPrecision(inputLogicalType);
+        int targetPrecision = LogicalTypeChecks.getPrecision(targetLogicalType);
+
+        if (inputLogicalType.is(LogicalTypeRoot.TIMESTAMP_WITHOUT_TIME_ZONE)
+                && targetLogicalType.is(LogicalTypeRoot.TIMESTAMP_WITHOUT_TIME_ZONE)) {
+            final TimestampKind inputTimestampKind = ((TimestampType) inputLogicalType).getKind();
+            final TimestampKind targetTimestampKind = ((TimestampType) targetLogicalType).getKind();
+            if (inputTimestampKind == TimestampKind.ROWTIME
+                    || inputTimestampKind == TimestampKind.PROCTIME
+                    || targetTimestampKind == TimestampKind.ROWTIME
+                    || targetTimestampKind == TimestampKind.PROCTIME) {
+                targetPrecision = 3;
```

Comment:
why is it necessary to change the target precision in this case? What happens if we don't apply this?

Reply:
we want `targetPrecision = 3` because everywhere in the code for performance reasons we treat those timestamps always as long, which means only milliseconds precision. If we allow more precision, then we can still make comparisons, etc. but any code that uses those timestamps must be checked and potentially, we would need to push down the pipeline the chosen precision, so the code needs how to process that if for example needs to print them like strings.

Reply:
thanks for the explanation, make sense for now.